stackrox/stackrox
# Replace ColorMapPF and cvssSeverityColor with accessible alternatives

## Description

### Problems

1. Color text causes accessibility problems like assumption of perfect color vision.

2. Inconsistent colors for severity throughout UI.

3. Color assignment to CVSS score sometimes contradicts severity.

### Analysis

src/constants/severityColors.js

1. `severityColorMapPF`

    * `CVSSScoreLabel` used only in `ObservedCVEsTable` component on Vulnerability Management Images side panel.
    * `PolicySeverityLabel` used only on Policies page and Policy overview.
    * violationTableColumnDescriptors used only on Violations page.

2. `vulnerabilitySeverityColorMapPF`

    * `VulnerabilitySeverityLabel` used only in 3 tables on Vulnerability Management Images side panel.

3. `cvssSeverityColorMap`

    * `CvesByCvssScore` used on Vulnerability Management overview pages for component, deployment, image, node

4. `cvssSeverityTextColorMap` ditto the preceding item

5. `cvssSeverityColorLegend` ditto the preceding item

### Solution

Replace `ReactNode` with `ReactElement` in return type of IconText components that I missed in #5696

1. `severityColorMapPF`

    * Replace `CVSSScoreLabel` with plain text, especially because the table column is sortable.
    * Replace `PolicySeverityLabel` with `PolicySeverityIconText` element that renders color icon and plain text. The colors and icons are consistent with most recent work on Vulnerabilities (preview).
    * Replace PatternFly `Label` element (ditto the preceding point).

2. `vulnerabilitySeverityColorMapPF`

    * Replace `VulnerabilitySeverityLabel` with `VulnerabilitySeverityIconText` element that renders color icon and plain text. The colors and icons are consistent with most recent work on Vulnerabilities (preview).

    By the way, `VulnerabilitySeverityLabel` has a separate meaning as a type in current work on Vulnerabilities.

3. `cvssSeverityColorMap`

    * Replace with `vulnSeverityIconColors` from src/constants/visuals/colors.ts file.

4. `cvssSeverityTextColorMap`

    * Replace color text with plain text.

5. `cvssSeverityColorLegend`

    * Replace color text with plain text and move refactored legend data structure into component.

## Checklist

## Testing Performed

### Integration testing

In ui/apps/platform folder:

1. `export CYPRESS_ROX_POSTGRES_DATASTORE=true`

2. `yarn cypress-open`

    * policies/policiesTable.test.js
    * violations/violations.test.js
    * vulnmanagement/images.test.js

### Manual testing

1. Visit /main/policy-management

    * classic severity colors: notice pale purple Critical in the last row
        ![policies_classic](https://user-images.githubusercontent.com/11862657/234339090-31797880-bdd4-4d0c-9574-0cddc037b3f3.png)

    * PatternFly severity colors: notice red Critical in the last row
        ![policies_PatternFly](https://user-images.githubusercontent.com/11862657/234340838-2ebfa7eb-ec3b-44e5-91c2-e464d670fc1b.png)

2. Click a policy

    * classic severity colors
        ![policy_classic](https://user-images.githubusercontent.com/11862657/234339149-ac7b2982-3f49-49e5-af2b-c096706c0935.png)

    * PatternFly severity colors
        ![policy_PatternFly](https://user-images.githubusercontent.com/11862657/234340896-64b47de9-89f5-474e-89d3-67dde7c13f28.png)

3. Visit /main/violations

    * classic severity colors
        ![violations_classic](https://user-images.githubusercontent.com/11862657/234339180-39d23151-c2e8-42a8-a50b-76797f06076c.png)

    * PatternFly severity colors
        ![violations_PatternFly](https://user-images.githubusercontent.com/11862657/234340774-6e3b5a64-ac2f-4e33-b925-dbabdbe27033.png)

4. Visit /main/vulnerability-management/images and click an image

    * classic severity colors
        ![image_classic](https://user-images.githubusercontent.com/11862657/234339221-1dd49730-bf4d-4efe-b1c0-eee3360e09a5.png)

    * PatternFly severity colors in light theme
        ![image_light_PatternFly](https://user-images.githubusercontent.com/11862657/234340696-3d7d673c-ae18-4abb-a5d8-68983d3c0146.png)

    * PatternFly severity colors in dark theme
        ![image_dark_PatternFly](https://user-images.githubusercontent.com/11862657/234340737-01dae19a-6024-4ae7-9cad-d60f55565744.png)
